### DISCUSSION TITLE
Step by step migrations and PRAGMA user_version

### DISCUSSION THREAD

**Discussion by CillianMyles** [simolus3/drift#2639] [Q&A]
From what I can see `Drift` only bumps the `PRAGMA user_version` (`schemaVersion` stored in SQLite) after successful migration:
https://github.com/simolus3/drift/blob/develop/drift/lib/src/runtime/executor/helpers/engines.dart#L457-L464

Since the entire step-by-step migration is considered a single `OnUpgrade (Future<void>)`, if some steps pass, but one fails, the `PRAGMA user_version` will not be changed, even though some steps of the migration were successful. 

If this is not properly handled, and the app is opened again, the full migration will be attempted again, which itself can fail, leaving the app and db in a broken/mis-matched state regarding `schemaVersion`, `PRAGMA` and the actual db schema (tables, indexes, etc.).

Imagine the simplified example:
- schema 1 is a table called `table` with a single TEXT column `column1`
- schema 2 adds another TEXT column `column2`
- schema 3 adds another TEXT column `column3`
- schema 4 does something which results in an Exception

```dart
  @override
  int get schemaVersion => 4;

  @override
  MigrationStrategy get migration {
    return MigrationStrategy(
      onCreate: (migrator) => migrator.createAll(),
      onUpgrade: (migrator, from, to) {
        return stepByStep(
          from1to2: (migrator, schema) async {
            await migrator.addColumn(schema.table, schema.column2);
          },
          from2to3: (migrator, schema) async {
            await migrator.addColumn(schema.table, schema.column3);
          },
          from3to4: (migrator, schema) async {
            // something which results in an Exception
          },
        );
      },
      beforeOpen: (details) async {
        // ...
      },
    );
  }
```

First time the app opens and migration from v1 → v4 is attempted:
- v1 → v2 ✅ 
- v2 → v3 ✅  
- v3 → v4 ❌ fails... so then `PRAGMA user_version` is never changed by `Drift`. It remains as `1` (not `3` which the schema actually has been migrated to).

Next time opening the app, the whole migration process is re-attempted:
- v1 → v2 ❌ fails... this is attempted at all because the `PRAGMA user_version` is still `1`. But when trying to add a column that already exists we get an `SqliteException` something like:
  ```
  SqliteException(1): while executing, duplicate column name: column2, SQL logic error (code 1)
  Causing statement: ALTER TABLE "table" ADD COLUMN "column2" TEXT;
  ```

Would it make sense for `Drift`, or failing that us as consumers, to bump the `PRAGMA user_version` upon completion of each successful step in a step-by-step migration?

We recently had this issue, and had to do a pseudo-flip-tables-migration when we detected this mismatch between `schema` and `PRAGMA` and migrations were re-attempted and failing continuously.

Am I missing something here?

**Comment by simolus3** [comment]
You're right - drift should change the `user_version` pragma after every migration step to avoid this problem. I'll fix this for the next version.
